naps/cores/cmv12k/s7_phy.py

```diff
@@ -0,0 +1,471 @@
+            if best_clock_match < clock_match:
+                best_clock_match = clock_match
+                best_clock_delay = clock_delay
+
+            print("[{:02d}] = {:02d}".format(clock_delay, clock_match))
+            self.ctrl_lane_delay_inc = 2
+
+        self.set_ctrl_delay(1, best_clock_delay)
+        return best_clock_delay, best_clock_match
+
+    @driver_method
+    def data_alignment(self):
+        num_lanes = 32
+
+        # determine the best bitslip as the one which matches over the most delays
+        best_bitslip = [0]*(num_lanes+1)
+        best_bitslip_matches = [0]*(num_lanes+1)
+        for bitslip in range(6):
+            match_count = [0]*(num_lanes+1)
+            self.data_lane_delay_reset = (1<<num_lanes)-1 # zero all lane delays
+            self.ctrl_lane_delay_reset = 1
+            for lane_delay in range(32):
+                all_matches = self.data_lane_match | (self.ctrl_lane_match << num_lanes)
+                for lane in range(num_lanes+1):
+                    match_count[lane] += 1 if all_matches & (1 << lane) else 0
+                self.data_lane_delay_inc = (1<<num_lanes)-1 # increment all delays
+                self.ctrl_lane_delay_inc = 1
+            for lane in range(num_lanes+1):
+                if best_bitslip_matches[lane] < match_count[lane]:
+                    best_bitslip_matches[lane] = match_count[lane]
+                    best_bitslip[lane] = bitslip
+            self.data_lane_bitslip = (1<<num_lanes)-1 # slip all bits
+            self.ctrl_lane_bitslip = 1
+
+        # apply the determined bitslip to all the channels
+        for bitslip in range(6):
+            apply_mask = 0
+            for lane in range(num_lanes):
+                if best_bitslip[lane] > 0:
+                    apply_mask |= (1 << lane)
+                    best_bitslip[lane] -= 1
+            self.data_lane_bitslip = apply_mask
+            if best_bitslip[num_lanes] > 0:
+                self.ctrl_lane_bitslip = 1
+                best_bitslip[num_lanes] -= 1
+
+        # measure the delay window for all the channels to determine the final delay
+        delay_window = [0]*(num_lanes+1)
+        self.data_lane_delay_reset = (1<<num_lanes)-1 # zero all lane delays
+        self.ctrl_lane_delay_reset = 1
+        for delay in range(32):
+            all_matches = self.data_lane_match | (self.ctrl_lane_match << num_lanes)
+            for lane in range(num_lanes+1):
+                delay_window[lane] |= (1 << delay) if all_matches & (1 << lane) else 0
+            self.data_lane_delay_inc = (1<<num_lanes)-1 # increment all delays
+            self.ctrl_lane_delay_inc = 1
+
+        best_delay = [0]*(num_lanes+1)
+        for lane, window in enumerate(delay_window):
+            if window == 0xFFFFFFFF: # all delays work
+                best_delay[lane] = 16 # so pick the middle
+            elif window & 0x80000000: # large delays work but small ones don't
```

Comment:
What about a case like this:
`window = 0b1111_0000_0000_0000_0000_0000_0000_1111`
This would select `lsb = 0`, so a delay of `16` but that obviously is not in the range that works.

Reply:
This logic has been reworked and this case is added